### ISSUE TITLE
Incorrect slot auction winner calculated (Kusama #82)

### ISSUE THREAD

**Issue by JelliedOwl** [paritytech/polkadot#7221]
I believe that the recently completed auction 82 on Kusama has demonstrated a bug in the winner selection calculation. https://kusama.subscan.io/auction/82

[ I should declare that I'm the holder of paraid 2268, which benefitted from the bug.  Hopefully no-one is going to argue that we should lose it, since we need to urgently renew the Kabocha parachain lease.  Bit.Country (who I think should have won) may or may not be unhappy with the outcome.   - But this is probably an on-chain governance discussion.]

# Why I think there's a bug
The winner was determind at this block: https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Frpc-kusama.luckyfriday.io#/explorer/query/17884265
<img width="681" alt="Screenshot 2023-05-12 at 23 35 21" src="https://github.com/paritytech/polkadot/assets/80860490/42b52cef-41ba-4103-ac63-305bcd2a96ee">

Auction block 3440 is kusama block number 17880720.

The final valid bid in the auction was by paraid 2096 at block 17877543 for 500KSM.  This should have been the winning bid, I believe, but my paraid (2268) has been awarded the lease.

# Potential causes for the bug

There were a few things unusual about the auction - one of which might have caused this issue.

- The auction opened with no bidders for the first few days.
- There was a lot of back-and-forth self-funded bids on paraids 2268 and 2096 (though I don't think this was the issue).
- After the final bid by 2096, I was running out of funds, so I created a crowdloan on the same paraid (I was considering getting someone to support the crowdloan and take the lead again - but we decided against it before adding the funds).
- I then tried to self bid again, which failed due to lack of funds caused by the 100KSM crowdloan creation lock (which I don't think is documented?)
- Someone came and added 1KSM to the crowdloan.

Initially, I thought there might have been some kind of wrap around issue resulting from the non-bid period at the start of the auction, but this doesn't seem to fit and I suspect the result is determined by using the Auction block index (3440 in this case) rather than one iteration though the action states, block by block.

*MOST LIKELY* I suspect the actual issue related to my paraid having several self bids and then a crowdloan on it.  Unfortunately, I didn't anticipate the bug so didn't grab a copy of the on-chain auction state before it was cleared at the end of the epoch.

**Comment by JelliedOwl** [comment]
Oh, there's a snapshot here from about a day before the auction ended.  Not sure how long Polkachu keeps them, though.
https://snapshots.polkachu.com/snapshots/kusama/kusama_17858266.tar.lz4

**Comment by burdges** [comment]
We do "candle actions" so that automated public bidders aka crowd loans aka smart contracts won't be too disadvantaged.  We determine auction ending times randomly using on-chain randomness produced by block VRFs, but this also requires the ending being retrospective, aka nobody knows if the end already happened while bidding.  See https://polkadot.network/blog/research-update-the-case-for-candle-auctions#thecaseforthecandleauction or https://papers.ssrn.com/sol3/papers.cfm?abstract_id=3846363&ref=cms.polkadot.network

**Comment by JelliedOwl** [comment]
Thanks @burdges, but I don't think it's that I'm misunderstanding how the auction works (though I never discount that as a possibility).

During the auction, it take 3600 snapshots of the auction state, 1 every 20 blocks.  At the end of the epoch after the auction completes, it picks a random index to choose one of those 3600 snapshots.

In this case, it selected snapshot 3440.  At that snapshot, I believe the auction state was:
paraid 2113 = 1KSM self-funded 
paraid 2268 = 400KSM self-funded (and 1KSM crowdloan, which shouldn't have an effect)
paraid 2096 = 500KSM self-funded 

On that state, paraid 2096 should have won.  But my paraid, 2268, was awarded the lease.

Visually, I believe this graph from parachains.info is a correct representation of the auction:
<img width="777" alt="Screenshot 2023-05-12 at 20 49 04" src="https://github.com/paritytech/polkadot/assets/80860490/76c87214-23db-41c8-9c35-1ebbec274755">

Obviously, I'm not eager to return the lease I shouldn't have been awarded, since I need to swap the Kabocha parachain to it pretty urgently, but I'd like to be certain that no-one is going to take it away the lease before I start doing that work.  And, indeed, I think the bug should be fixed so it doesn't happen again.

**Comment by bkchr** [comment]
Looking at the data:
```
[
  null
  null
  null
  null
  null
  null
  null
  [
    EuLcF5dhnXJNjcToTjAXwGVERdoUNEmrDGrumahKt3r6f2p
    2,268
    401,000,000,000,000
  ]
  null
  null
  null
  null
  null
  null
  [
    F37GR5PvHiNj8YWVegr8r2nMosBDtiPVdUrfBSAMZg3e5HJ
    2,113
    1,000,000,000,000
  ]
  null
  null
  null
  null
  null
  null
  null
  null
  null
  null
  [
    Ff9JGfDpZ9TjLh7C2NXSUNAd3rAFNFqefxWE35vjCFyXKr4
    2,096
    500,000,000,000,000
  ]
  null
  null
  null
  null
  null
  null
  null
  null
  null
  null
]
```

The bare data shows what you would expect. However, the magic is hidden here:
https://github.com/paritytech/polkadot/blob/159231d7f80187da394a143f0f83ec940d5da5cc/runtime/common/src/auctions.rs#L634-L634

The length of a range is defined as:
https://github.com/paritytech/polkadot/blob/159231d7f80187da394a143f0f83ec940d5da5cc/runtime/common/slot_range_helper/src/lib.rs#L151-L156

Your parachain 2268 seems to have applied for the full lease, `ZeroSeven` which should have a length of `8`.

Parachain 2097 has applied for the lease `ThreeSeven` which should have a length of `5`.

`8 * 401K > 5 * 500K`.

I'm no expert on this topic, but it looks correct to me?

The graph from parachains.info should be corrected to show the correct data.

**Comment by shawntabrizi** [comment]
Not a bug, and not "incorrectly calculated".

This is a preference of the code to value bids for longer lease periods over those with shorter lease periods.

Not saying whether this makes sense or not, the algorithm and behavior is designed by Web3 foundation, but as @bkchr says above, basically `8 * 401K > 5 * 500K`. The way that a short lease period bid would win, is if there was also a bid for the unfilled lease periods from another person, which would add to the total value locked. But, it does not seem that parachains have coordinated enough yet to do these kinds of bids together.

I.e. Parachain A bids for lease periods 0-3 and Parachain B bids for lease periods 4-7, and together, their bid values work together to win over someone who bid less for 0 - 7.

**Comment by JelliedOwl** [comment]
Thank you very much @bkchr, @shawntabrizi.  That makes sense now - at least in terms of it not being a bug.  Feel free to close.

It's a really tricky situation to present on a graph - and I suspect both Bit.Country and I misunderstood it (though they've secured their renewal now anyway). And, I have just tested that I can make a tiny bid for 7-7 in the current auction, even though it pretty much can't win (unless someone jumps in and makes a large bid for 0-6 which is highly unlikely), so the bidding isn't blocked by the larger overall bid being higher - which I thought it might be.

**Comment by JelliedOwl** [comment]
It does actually have a pop-up that somewhat explains it, which I missed:
<img width="515" alt="Screenshot 2023-05-22 at 16 25 08" src="https://github.com/paritytech/polkadot/assets/80860490/46a69591-9ccb-47ec-887b-8f998494e02a">

**Comment by bkchr** [comment]
Yeah, they also explained it on twitter: https://twitter.com/parachains/status/1657123259916120066?cxt=HHwWhMCzhee0pP8tAAAA

I think it would better to directly show the valuation in the graph!

Going to close the issue!
